File: mailblastr-php/src/Resources/Resource.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Base class for API resources.
 */
abstract class Resource
{
    public function __construct(protected readonly Client $client)
    {
    }

    /**
     * Build a `?limit=&after=&before=` cursor-pagination query string from
     * params (plus any extra whitelisted keys).
     */
    protected function paginationQuery(array $params, array $extraKeys = []): string
    {
        $keys = array_merge(['limit', 'after', 'before'], $extraKeys);
        return $this->client->query(array_intersect_key($params, array_flip($keys)));
    }
}

File: mailblastr-java/src/main/java/com/mailblastr/resources/Resource.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.http.ApiClient;
import com.mailblastr.http.Query;
import com.mailblastr.http.Urls;

/** Shared plumbing for all resource services. */
public abstract class Resource {
    protected final ApiClient api;

    protected Resource(ApiClient api) { this.api = api; }

    /** Percent-encode a path segment (ids, emails, …). */
    protected static String enc(String segment) { return Urls.encode(segment); }

    /** Build a {@code ?limit=&after=&before=} query string from pagination params. */
    protected static String paginate(ListParams params) {
        Query q = new Query();
        if (params != null) params.applyTo(q);
        return q.toString();
    }
}
